drivers:platform:stm32:Support for PWM with Interrupt

jansunil opened the pull request:
Added support for PWM with timer interrupt in stm32

## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [ ] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [ ] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [ ] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [ ] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

buha commented on include/no_os_pwm.h:
```diff
@@ -82,6 +82,10 @@ struct no_os_pwm_init_param {
 	enum no_os_pwm_polarity polarity;
 	/** PWM gpio pin init param*/
 	struct no_os_gpio_init_param *pwm_gpio;
+	/* Timer IRQ ID */
```
In no_os_pwm.h we shouldn't care about the specific peripheral that is used to create the pulse width modulated signal. The peripheral could be:
* dedicated PWM peripheral
* Timer with PWM feature (as in your stm32 implementation)
* Timer+Interrupts+GPIO manually handled to create the signal

So the point is, you will need to rename tim_callback to pwm_callback and don't mention anything about a Timer because the implementation on another platform could use a dedicated PWM peripheral for example, and not a Timer...

jansunil replied:
Got your point here. 
I've re-worded the struct members/comments by removing the term 'timer'

jansunil commented:
V1: Removed the term "timer" from the no-os PWM driver
Reworded comments and variable names to remove the "timer" prefix

CiprianRegus commented on drivers/platform/stm32/stm32_pwm.c:
```diff
@@ -170,6 +172,35 @@ static int32_t stm32_init_timer(struct stm32_pwm_desc *desc,
 	desc->timer_autoreload = sparam->timer_autoreload;
 
+	if (param->pwm_callback) {
+		ret = lf256fifo_init(&sparam->rx_fifo);
+		if (ret < 0)
+			goto error;
```
Why is this FIFO required?

jansunil replied:
This is actually not required. Removed .

jansunil commented on drivers/platform/stm32/stm32_pwm.c:
```diff
@@ -170,6 +172,35 @@ static int32_t stm32_init_timer(struct stm32_pwm_desc *desc,
+
+		ret = no_os_irq_enable(desc->nvic_tim, param->irq_id);
+		if (ret < 0)
+			goto error;
```
Added steps for error handling. Does it look ok now?

jansunil commented on drivers/platform/stm32/stm32_pwm.c:
```diff
@@ -42,6 +42,8 @@
 #include "stm32_gpio.h"
 #include "stm32_pwm.h"
+#include "stm32_irq.h"
+#include "no_os_lf256fifo.h"
```
No, removed

jansunil commented:
V2: Code updates after review comments
1) Removed rx_fifo and related functionalities as it is not required
2) Added error handling steps in stm32_timer_init() function
3) Added unregister callback and irq remove in the stm32_pwm_remove() function

CiprianRegus commented on drivers/platform/stm32/stm32_pwm.c:
```diff
@@ -346,6 +378,19 @@ int32_t stm32_pwm_remove(struct no_os_pwm_desc *desc)
 	if (ret)
 		return ret;
 
+	if (desc->irq_id) {
+		ret = no_os_irq_unregister_callback(extra->nvic_tim, desc->irq_id,
+						    &extra->timer_callback);
+		if (ret) {
+			return ret;
+		}
+
+		ret = no_os_irq_ctrl_remove(extra->nvic_tim);
+		if (ret) {
+			return ret;
+		}
+	}
```
you also have to call `no_os_irq_ctrl_remove()`

jansunil commented:
V3: Added no_os_irq_disable() in the stm32_pwm_remove()
